src/main/java/uk/gov/hmcts/ccd/domain/service/aggregated/MergeDataToSearchResultOperation.java

```diff
@@ -24,32 +27,49 @@ public MergeDataToSearchResultOperation(final UIDefinitionRepository uiDefinitio
         this.uiDefinitionRepository = uiDefinitionRepository;
     }
 
-    public SearchResultView execute(final CaseType caseType, final List<CaseDetails> caseDetails,final String view) {
+    public SearchResultView execute(final CaseType caseType, final List<CaseDetails> caseDetails, final String view, final String resultError) {
         final SearchResult searchResult = getSearchResult(caseType, view);
         final List<SearchResultViewColumn> viewColumns = Arrays.stream(searchResult.getFields())
             .flatMap(searchResultField -> caseType.getCaseFields().stream()
-            .filter(caseField -> caseField.getId().equals(searchResultField.getCaseFieldId()))
-            .map(caseField ->  new SearchResultViewColumn(
-                searchResultField.getCaseFieldId(),
-                caseField.getFieldType(),
-                searchResultField.getLabel(),
-                searchResultField.getDisplayOrder(),
-                searchResultField.isMetadata())
-             ))
-            .collect(Collectors.toList());
+                .filter(caseField -> caseField.getId().equals(searchResultField.getCaseFieldId()))
+                .map(caseField -> new SearchResultViewColumn(
+                    searchResultField.getCaseFieldId(),
+                    caseField.getFieldType(),
+                    searchResultField.getLabel(),
+                    searchResultField.getDisplayOrder(),
+                    searchResultField.isMetadata()))
+                )
+             .collect(Collectors.toList());
 
         final List<SearchResultViewItem> viewItems = caseDetails.stream()
-            .map(caseData -> new SearchResultViewItem(caseData.getReference().toString(), caseData.getCaseDataAndMetadata()))
+            .map(caseData -> buildSearchResultViewItem(caseData))
             .collect(Collectors.toList());
-        return new SearchResultView(viewColumns, viewItems);
+        return new SearchResultView(viewColumns, viewItems, resultError);
+    }
+
+    private SearchResultViewItem buildSearchResultViewItem(CaseDetails caseData) {
+        return new SearchResultViewItem(hasCaseReference(caseData) ? getCaseReference(caseData) : getDraftReference(caseData),
+                                        caseData.getCaseDataAndMetadata());
+    }
+
+    private String getCaseReference(CaseDetails caseData) {
+        return caseData.getReference().toString();
+    }
+
+    private String getDraftReference(CaseDetails caseData) {
+        return String.format(DRAFT_ID, caseData.getId());
+    }
+
+    private boolean hasCaseReference(CaseDetails caseData) {
+        return caseData.getReference() != null;
```

Comment:
The way to do this in Java 8 is to have the methods return Optional. We definitely should change getReference() to return an optional now. This is very important. If we don't do this, we risk many subtle bugs and null pointer exceptions in the future whenever we'll retrieve the reference

Reply:
But then do you want to change Long reference to Optional<Long>? How would jackson ser/deser that. Also should we have it as a separate tech debt story? Btw, I do not think we should change it.

Reply:
if we don't change it, and in the future we call getReference() forgetting to do the null check, how can we prevent null pointer exceptions?